sys/include/crypto/dragon.h

```diff
@@ -0,0 +1,364 @@
+    sbox1[(x >> 16) & 0xFF] ^	\
+    sbox1[(x >> 24) & 0xFF]
+
+#define G2(x)			\
+    sbox1[x & 0xFF] ^		\
+    sbox2[(x >> 8) & 0xFF]  ^	\
+    sbox1[(x >> 16) & 0xFF] ^	\
+    sbox1[(x >> 24) & 0xFF]
+
+#define G3(x)			\
+    sbox1[x & 0xFF] ^		\
+    sbox1[(x >> 8) & 0xFF]  ^	\
+    sbox2[(x >> 16) & 0xFF] ^	\
+    sbox1[(x >> 24) & 0xFF]
+
+#define H1(x)			\
+    sbox1[x & 0xFF] ^		\
+    sbox2[(x >> 8) & 0xFF]  ^	\
+    sbox2[(x >> 16) & 0xFF] ^	\
+    sbox2[(x >> 24) & 0xFF]
+
+#define H2(x)			\
+    sbox2[x & 0xFF] ^		\
+    sbox1[(x >> 8) & 0xFF]  ^	\
+    sbox2[(x >> 16) & 0xFF] ^	\
+    sbox2[(x >> 24) & 0xFF]
+
+#define H3(x)			\
+    sbox2[x & 0xFF] ^		\
+    sbox2[(x >> 8) & 0xFF]  ^	\
+    sbox1[(x >> 16) & 0xFF] ^	\
+    sbox2[(x >> 24) & 0xFF]
+
+
+/**
+ * @brief  Dragon Key setup
+ *
+ * @param[in]   ctx     Dragon context
+ * @param[in]   key     Key to setup the ctx
+ * @param[in]   keysize Size of the key
+ */
+void dragon_keysetup(dragon_ctx *ctx, const uint8_t *key, uint32_t keysize);
+
+/**
+ * @brief  Dragon IV setup
+ *
+ * @param[in]   ctx     Dragon context
+ * @param[in]   iv      IV to setup the ctx
+ */
+void dragon_ivsetup(dragon_ctx *ctx, const uint8_t *iv);
+
+/**
+ * @brief  Generate "blocks" number of keystream blocks
+ *
+ * @param[in]   ctx         Dragon context
+ * @param[out]  keystream   Pre-allocated array containing 8*(blocks)
+ *                          bytes of memory
+ * @param[in]   blocks      Number of keystream blocks to produce
+ */
+void dragon_keystream_blocks(dragon_ctx *ctx, uint8_t *keystream, uint32_t blocks);
+
+/**
+ * @brief  Generate "length" number of keystream bytes
+ *
+ * @param[in]   ctx         Dragon context
+ * @param[out]  keystream   Pre-allocated array containing (msglen)
+ *                          bytes of memory
+ * @param[in]   length      Number of keystream bytes to produce
+ */
+void dragon_keystream_bytes(dragon_ctx *ctx, uint8_t *keystream, uint32_t length);
+
+/**
+ * @brief  Encrypt or decrypt a "blocks" number of blocks
+ *
+ * @param[in]   ctx         Dragon context
+ * @param[in]   input       (Plain/Cipher)text blocks for (en/de)crypting
+ * @param[out]  output      Pre-allocated array for (cipher/plain)text blocks
+ *                          consisting of 8*(blocks) bytes
+ * @param[in]   blocks      Number of blocks to (en/de)crypt
+ */
+void dragon_process_blocks(dragon_ctx *ctx, const uint8_t *input, uint8_t *output, uint32_t blocks);
+
+/**
+ * @brief  Encrypt or decrypt "msglen" number of message bytes
+ *
+ * @param[in]   ctx         Dragon context
+ * @param[in]   input       (Plain/Cipher)text for (en/de)crypting
+ * @param[out]  output      (Cipher/Plain)text for (en/de)crypting
+ * @param[in]   msglen      Number of bytes to (en/de)crypt
+ */
+void dragon_process_bytes(dragon_ctx *ctx, const uint8_t *input, uint8_t *output, uint32_t msglen);
+
+/**
+ * @brief The Dragon sbox definitions
+ */
+static const uint32_t sbox1[256] = {
```

Comment:
No i did not took it like that. i think the discussion is a little bit deadend. So on one side you can say - "hey put all macros aways from the c code into headers, so the c code looks cleaner", others say "hey i dont want to open 4 files to read the code and it defs please only list the pub functions in header files and pack all the rest into the c file". With crypto code most of these macro defines should never be touched and are are wired kind of way to "segment the algorithm code", beatifull no, working and understandable yes.

ill work to fix some stuff - some ciphers are also more or less define free. i am currently implementing ed25519 and curve25519 and creating a python script to move functions to defines and vice versa.

in my personal opinion the best idea would be to leave it as it is (maybe some more fancy naming for the defines) and simply having a clean cipher-api.h file that only includes the ~3 public functions declarations. cipher.h is only included by cipher.c -> clean. what you think ahout that?